### PR TITLE
Update internal code for at::_lu_with_info

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56613 Update internal code for torch.linalg.solve
* **#56612 Update internal code for at::_lu_with_info**
* #56611 Update internal code for torch.lu_solve

The goal of this refactoring is to make the `torch.linalg.solve`
to be a composition of calls to `lu_stub` and `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Replaced `lu_with_info_{cpu, cuda}` with one function that calls
to `lu_stub`.
Split MAGMA-based `apply_lu` into `apply_lu_looped_magma`
and `apply_lu_batched_magma`. This simplifies the future switch to
cuSOLVER and cuBLAS libraries.

Differential Revision: [D28248756](https://our.internmc.facebook.com/intern/diff/D28248756)

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp [hunk 1]
```diff
@@ -766,6 +766,55 @@ void triangular_solve_kernel(Tensor& A, Tensor& B, Tensor& infos, bool upper, bo
...
+  auto leading_dimension = std::max<int64_t>(1, m);
+
+  for (const auto i : c10::irange(batch_size)) {
+    scalar_t* input_working_ptr = &input_data[i * input_matrix_stride];
+    int* pivots_working_ptr = &pivots_data[i * pivots_stride];
+    int* infos_working_ptr = &infos_data[i];
+    lapackLu<scalar_t>(m, n, input_working_ptr, leading_dimension, pivots_working_ptr, infos_working_ptr);
+  }
+#endif
+}
```
Comment (nikitaved) [inline aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:809]: TODO: for the CPU we could use `batch_iterator_with_broadcasting`. But we have to be careful not to materialize redundant copies prior to calling to the CPU dispatch.

**Review by nikitaved — COMMENTED** [review]
LGTM! Thank you, @IvanYashchuk !

**Review by mruberry — APPROVED** [review]
Stamped! Thanks @IvanYashchuk, and thanks for reviewing @nikitaved

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28248756).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28248756).

**Pull request closed without merging**